### PR TITLE
Enabling more control over when checkpoints are saved and the number of times loss updates are performed.

### PR DESCRIPTION
**Pull request by Lucaweihs**
Updates from the title as well as some improvements to the ObjectNav code (e.g. @mattdeitke's changes allowing for randomized materials in training).

### CODE REVIEW

**Review by jordis-ai2 — APPROVED** [review]
Just some very minor comments (some of them for myself). Other than that, LGTM.

**Comment by jordis-ai2** [comment]
Since we first merged the changes in multinode, please let me know if you want me to resolve the conflicts here.

File: allenact/algorithms/onpolicy_sync/engine.py [hunk 1]
```diff
@@ -930,12 +935,17 @@ def update(self, rollouts: RolloutStorage):
 
                 info: Dict[str, float] = {}
 
+                current_pipeline_stage = self.training_pipeline.current_stage
                 total_loss: Optional[torch.Tensor] = None
                 for loss_name in self.training_pipeline.current_stage_losses:
-                    loss, loss_weight = (
+                    loss, loss_weight, loss_update_repeats = (
```
[the remark this answers is not in the document]
Comment (Lucaweihs) [reply allenact/algorithms/onpolicy_sync/engine.py:1002]: Yah, I realized I needed this fine control to be able to do this in the ADVISOR when training something with GAIL where the GAIL discriminator loss needs to be updated more than other losses. In any case, I've merged in your multi-node updates and I think it all works correctly now (we can specify `update_repeats` at the loss, pipeline stage, or training pipeline level).

File: allenact/algorithms/onpolicy_sync/engine.py [hunk 2]
```diff
@@ -1168,17 +1178,60 @@ def run_pipeline(self, rollouts: RolloutStorage):
...
+                and (  # Might happen if the `save_interval` was hit
+                    not already_saved_checkpoint
+                )
+                and pipeline_stage_changed
+                and (  # Don't save at start
+                    self.training_pipeline.current_stage_index != 0
+                )
+                and (self.save_ckpt_after_every_pipeline_stage or training_is_complete)
+            ):
+                # TODO: For the moment these lines are repeated verbatim below, this should
```
[the remark this answers is not in the document]
Comment (Lucaweihs) [reply allenact/algorithms/onpolicy_sync/engine.py:1328]: I'll do this, not sure why I left it as a TODO.

File: allenact/algorithms/onpolicy_sync/vector_sampled_tasks.py [hunk 3]
```diff
@@ -82,9 +83,20 @@ def __exit__(self, type, value, traceback):
         signal.signal(signal.SIGINT, self.old_int_handler)
         signal.signal(signal.SIGTERM, self.old_term_handler)
         if self.term_signal_received:
-            self.old_term_handler(*self.term_signal_received)
+            if callable(self.old_term_handler):
```
[the remark this answers is not in the document]
Comment (Lucaweihs) [reply allenact/algorithms/onpolicy_sync/vector_sampled_tasks.py:86]: Honestly I'm not entirely sure why this is necessary either but I was finding that I sometimes received error messages about `self.old_term_handler` not being callable. I'll add the following comment:
```python
            # For some reason there appear to be cases where the original termination
            # handler is not callable. It is unclear to me exactly why this is the case
            # but here we add a guard to double check that the handler is callable and,
            # if it's not, we re-send the termination signal to the process and let
            # the python internals handle it (note that we've already reset the termination
            # handler to what it was originaly above in the signal.signal(...) code).
```

File: allenact/base_abstractions/sensor.py [hunk 4]
```diff
@@ -393,27 +393,3 @@ def query_expert(
         return task.query_expert(
             **self.expert_args, expert_sensor_group_name=expert_sensor_group_name
         )
-
-
-class VisionSensor:
```
[the remark this answers is not in the document]
Comment (Lucaweihs) [reply allenact/base_abstractions/sensor.py:398]: I know I was the one originally arguing for forwarding users in this way but it ended up causing more problems than it was worth (Kiana/I independently ran into problems where doing this type of forwarding resulted in weird issues, for me this was when some old code was doing something like `isinstance(obj, RGBSensor)` and then failing subtly because `RGBSensor` was one of these stubs rather than the true `RGBSensor`.

File: projects/pointnav_baselines/experiments/pointnav_thor_base.py [hunk 5]
```diff
@@ -54,7 +54,6 @@ def __init__(self):
             height=self.CAMERA_HEIGHT,
             continuousMode=True,
             applyActionNoise=self.STOCHASTIC,
```
[the remark this answers is not in the document]
Comment (Lucaweihs) [reply projects/pointnav_baselines/experiments/pointnav_thor_base.py:56]: Double checked just now and it seems to work as I expect (agent rotation/movement is stochastic).

File: allenact_plugins/robothor_plugin/robothor_task_samplers.py [hunk 6]
```diff
@@ -451,6 +453,15 @@ def next_task(self, force_advance_scene: bool = False) -> Optional[ObjectNavTask
...
 
+        # only randomize materials in train scenes
+        if self.randomize_materials_in_training:
+            if (
+                "Train" in scene
+                or int(scene.replace("FloorPlan", "").replace("_physics", "")) % 100
+                < 21
+            ):
+                self.env.controller.step(action="RandomizeMaterials")
+
```
[the remark this answers is not in the document]
Comment (Lucaweihs) [reply allenact_plugins/robothor_plugin/robothor_task_samplers.py:466]: 👍 done

File: projects/objectnav_baselines/experiments/objectnav_thor_base.py [hunk 7]
```diff
@@ -49,16 +49,21 @@ class ObjectNavThorBaseConfig(ObjectNavBaseConfig, ABC):
...
+    AGENT_MODE = "default"
+
     TARGET_TYPES: Optional[Sequence[str]] = None
 
     THOR_COMMIT_ID: Optional[str] = None
 
     def __init__(
         self,
         num_train_processes: Optional[int] = None,
+        num_test_processes: Optional[int] = None,
```
[the remark this answers is not in the document]
Comment (Lucaweihs) [reply projects/objectnav_baselines/experiments/objectnav_thor_base.py:63]: Do more changes need to made for this or should it work now? I thought it would be fine as test only ever runs on a single machine?

File: allenact/algorithms/onpolicy_sync/vector_sampled_tasks.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (jordis-ai2) [reply allenact/algorithms/onpolicy_sync/vector_sampled_tasks.py:92]: Thanks for the note. I'm not 100% sure if this covers all remaining failure cases of this kind, but even if we just get a bit closer, I guess it's fine :)

File: projects/objectnav_baselines/experiments/objectnav_thor_base.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (jordis-ai2) [reply projects/objectnav_baselines/experiments/objectnav_thor_base.py:63]: My doubt was actually about the new args to init, but I then realized you already have defaults for each. Never mind :)

**Review by jordis-ai2 — APPROVED** [review]
LGTM. Feel free to merge!
